iwyu.cc

```diff
@@ -3000,8 +2983,29 @@ class InstantiatedTemplateVisitor
   bool TraverseTemplateSpecializationTypeHelper(
       const clang::TemplateSpecializationType* type) {
     if (CanIgnoreCurrentASTNode())  return true;
-    if (CanForwardDeclareType(current_ast_node()))
-      current_ast_node()->set_in_forward_declare_context(true);
+
+    // Skip the template traversal if this occurrence of the template name is
+    // a class qualifier for an out of line method and not an actual type
+    // instantiation. Note also that in this case the method decl will not
+    // have any dependent types and so will be treated as non-templated code.
+    //
+    // Class<int>::method() {
+    // |-Type---^
+    // |-NNS------^
+    // |-CXXMethodDecl--^
+    //
+    ASTNode* const ast_node = current_ast_node();
+    if (const auto* nns = ast_node->GetParentAs<NestedNameSpecifier>()) {
```

Comment:
One some level I feel that maybe this code should move into iwyu_ast_util.cc, but I have a really hard time coming up with a name for it, so maybe it should just stay :)

Reply:
Yeah, I think so too. Not clear this functionality can be so useful to deserve its own function anyway. And it can always be changed if it turns out otherwise.